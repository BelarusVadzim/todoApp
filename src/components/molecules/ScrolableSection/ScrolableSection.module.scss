@import "src/variables.scss";

.scrolableSection {
  position: relative;
  max-height: $scrolableSection-max-height;
  overflow: auto;
}

.scrolableSection::-webkit-scrollbar {
  display: none;
}


