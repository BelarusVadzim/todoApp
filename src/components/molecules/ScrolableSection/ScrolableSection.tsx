import React from 'react';
import style from './ScrolableSection.module.scss';
import { arrayMoveImmutable } from 'array-move';
import { SortEvent, SortEventWithTag } from 'react-sortable-hoc';
import { useTodoStateService } from 'hooks';
import { filterTypes } from 'constants/filterTypes';
import { SortableSection } from '../SortableSection';

type SortEndArg = { oldIndex: number; newIndex: number };

const ScrolableSection: React.FC = () => {
  const { todos, dispatchTodoListChanged, todoFilter } = useTodoStateService();

  const onSortEnd = ({ oldIndex, newIndex }: SortEndArg) => {
    const todoList = arrayMoveImmutable(todos, oldIndex, newIndex);
    dispatchTodoListChanged(todoList);
  };

  const checkIsElementDraggable = (element: Element) => {
    return !!element.getAttribute('data-draggable');
  };
  const shouldCancelDrag = (event: SortEvent | SortEventWithTag) =>
    todoFilter === filterTypes.Active ||
    todoFilter === filterTypes.Completed ||
    todos.length < 2 ||
    !checkIsElementDraggable(event.target as Element);

  return (
    <div className={style.scrolableSection}>
      <SortableSection
        onSortEnd={onSortEnd}
        shouldCancelStart={shouldCancelDrag}
      />
    </div>
  );
};

export default ScrolableSection;
