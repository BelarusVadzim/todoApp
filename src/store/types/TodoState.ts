import { TodoNote } from 'types';

export type TodoState = {
  todos: TodoNote[];
  filter?: string;
  pending?: boolean;
};
