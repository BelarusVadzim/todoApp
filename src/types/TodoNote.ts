export type TodoNote = {
  id?: number;
  done?: boolean;
  text: string;
};
