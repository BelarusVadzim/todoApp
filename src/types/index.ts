export * from './PropsWithClassName';
export * from './TodoNote';
