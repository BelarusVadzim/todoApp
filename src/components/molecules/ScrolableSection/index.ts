import ScrolableSection from './ScrolableSection';

export { ScrolableSection };
