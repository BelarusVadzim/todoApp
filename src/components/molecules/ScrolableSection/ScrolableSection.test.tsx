import { fireEvent, render, screen } from '@testing-library/react';
import ScrolableSection from './ScrolableSection';
import { TodoNote } from 'types';

const dispatchTodoListChanged = jest.fn();
let shouldCancelStartValue: boolean;
const todo1 = { text: 'todo1' };
const todo2 = { text: 'todo2' };
const todos: TodoNote[] = [ todo1, todo2];
const sortedTodos: TodoNote[] = [ todo2, todo1 ];
const todoFilter = 'todoFilter';

jest.mock('./ScrolableSection.module.scss', () => ({
  scrolableSection: 'scrolableSectionStyle',
}));

jest.mock('hooks', () => ({
  useTodoStateService: () => ({
    dispatchTodoListChanged,
    todos,
  }),
  useAppStateService: () => ({ todoFilter }),
}));

jest.mock('array-move', () => ({
  arrayMoveImmutable: jest.fn(() => sortedTodos),
}));

jest.mock('../SortableSection', () => ({
  SortableSection: jest.fn(({ onSortEnd, shouldCancelStart }) =>  {
    const onClickHandler = () => {
      shouldCancelStartValue = shouldCancelStart({ 
        target: { 
          getAttribute: () => true,
        }, 
      });
    };
    
    return (<div data-testid="sortable-section" onDrop={onSortEnd} onClick={onClickHandler} />);
  }),
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('<ScrolableSection />', () => {
  it('should render properly', () => {
    const component = render(<ScrolableSection />);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('when slected element is draggable shouldCancelStart should be falsy', () => {
    render(<ScrolableSection />);

    const inputElement = screen.getByTestId('sortable-section');

    fireEvent.click(inputElement);

    expect(shouldCancelStartValue).toBeFalsy();
  });

  it('when sorting finished should call expected functions with expected values', () => {
    render(<ScrolableSection />);

    const inputElement = screen.getByTestId('sortable-section');

    fireEvent.drop(inputElement);

    expect(dispatchTodoListChanged).toBeCalledTimes(1);
    expect(dispatchTodoListChanged).toBeCalledWith(sortedTodos);
  });
});
