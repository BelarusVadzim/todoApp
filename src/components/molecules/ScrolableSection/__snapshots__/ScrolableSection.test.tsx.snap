// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScrolableSection /> should render properly 1`] = `
<DocumentFragment>
  <div
    class="scrolableSectionStyle"
  >
    <div
      data-testid="sortable-section"
    />
  </div>
</DocumentFragment>
`;
